### DISCUSSION TITLE
InWidth &OutWidth length of HLS streams.

### DISCUSSION THREAD

**Discussion by shashwat1198** [Xilinx/finn#501] [Q&A]
Hello,

I am trying to build a convolutional model using FINN. For some configurations of the model I get this error quite often: 

AssertionError: OutWidth is bigger than InWidth and is not divisible by it. Please adjust PE and SIMD values so that OutWidth % InWidth = 0

On reading about this error I found you have the option to manually set your PE and SIMD values in the automatic config build, where the PE and SIMD values should be such that these conditions for each layer are satisfied : 
Input Channels % SIMD == 0 & Output Channels % PE == 0.

For example, if one of the model layer looks like this :  
qnn.QuantConv2d(in_channels = 40,out_channels=80, kernel_size=2,weight_bit_width=2,bias=False),

Then I would assume that setting the PE & SIMD values to 8 & 10 should not cause any issues as the above conditions are satisfied (40%10=0 & 80%8=0).
But this does not seem to work and I see the above mentioned assertion error. 

On the other hand if I use PE & SIMD values as 8 & 8 (40%8=0 & 80%8=0) then the hardware build is successful and I see no errors.
So I think that Outwidth and Inwidth must be a multiple of one another.

This is what I wanted to ask : How would I go about computing the InWidth and OutWidth of the HLS streams corresponding to each layer so that I can then decide the PE & SIMD values accordingly?

If someone is familiar with this, I would very much appreciate your insight. Any pointers would be very helpful.
Shashwat 
@maltanar

**Comment by Benjamin-Laf** [comment]
Hello @shashwat1198,

I have the same error in the build IP with an mvau = 30 in my python configuration file. I have this return:
AssertionError: InWidth is bigger than OutWidth and is not divisible by it.
            Please adjust PE and SIMD values so that InWidth % OutWidth = 0

When I try with mvau = 40 I don't have this problem but I don't inderstand the problem appears after the estimation build (steps 1 to steps 7).
I work on because I want to change the folding of my layers and experiment this option for the throughput optimisation.

Finally, I have two questions:

_ Have you succeed to pass althrough this problem with an other less value for mvau or not ?
_ How are you change manually the values of PE and SIMD for each nodes ? (Only if you say the method for this)

I hope you understand my poor english and thanks for your future answer.
